src/vintch_model/subunit_model.py

```diff
@@ -0,0 +1,178 @@
+    def __call__(self, *args, **kwds):
+        return self.forward(*args, **kwds)
+
+    @property
+    def kernels(self):
+        return self._kernels
+
+    @kernels.setter
+    def kernels(self, new_kernels: Any):
+        """Setter for kernel filters."""
+        assert (
+            new_kernels.shape == self._kernels.shape
+        ), f"Expected kernel shape to be {self._kernels.shape}, got {new_kernels.shape} instead."
+        self.backend_class.check_input_type(new_kernels)
+        self._kernels = new_kernels
+
+    @property
+    def pooling_weights(self):
+        return self._pooling_weights
+
+    @pooling_weights.setter
+    def pooling_weights(self, weights: Any):
+        """Setter for pooling weights."""
+        assert (
+            weights.shape == self._pooling_weights.shape
+        ), f"Expected pooling weights shape to be {self._pooling_weights.shape}, got {weights.shape} instead."
+        self.backend_class.check_input_type(weights)
+        self._pooling_weights = weights
+
+    @property
+    def pooling_biases(self):
+        return self._pooling_biases
+
+    @pooling_biases.setter
+    def pooling_biases(self, biases: Any):
+        """Setter for pooling biases."""
+        assert (
+            biases.shape == self._pooling_biases.shape
+        ), f"Expected pooling biases shape to be {self._pooling_biases.shape}, got {biases.shape} instead."
+        self.backend_class.check_input_type(biases)
+        self._pooling_biases = biases
+
+    def forward(self, x):
+        """
+        Forward pass through the whole model.
+        """
+        self.backend_class.check_input_type(x)
+        # ensure the input is 5D and the second dimension is 1 (grayscale)
+        if x.ndim != 5 or x.shape[1] != 1:
+            raise ValueError(
+                f"Input must be a 5D tensor with shape [batch, {self.n_channels}, time, height, width], got {x.shape} instead."
+            )
+        generator_signal = self.generator_forward(x)  # [batch, time]
+        out = self.nonlinearity_out.transform(generator_signal)  # [batch, time]
+        return out
+
+    def generator_forward(self, x):
+        convolved = self.convolve(
+            x, self.kernels
+        )  # [batch, channels, time, height, width]
+        activated = self.nonlinearity_pass(
+            convolved, self.nonlinearities_chn
+        )  # [batch, channels, time, height, width]
+        pooled = self.weighted_pooling(
+            activated, self.pooling_weights, biases=self.pooling_biases
+        )  # [batch, channels, time]
+        output = pooled.sum(axis=1)  # [batch, time]
+        return output
+
+    def convolve(self, x, kernel):
+        return self.backend_class.convolve(x, kernel)
+
+    def nonlinearity_pass(self, x, nonlinearities):
+        transformed = [
+            nonlinearity.transform(x[:, c])
+            for c, nonlinearity in enumerate(nonlinearities)
+        ]
+        return self.backend_class.lib.stack(transformed, axis=1)
```

Comment:
I think stack can also be handled by [einops.pack](https://einops.rocks/4-pack-and-unpack/#stacking-data-layers)

Reply:
which I'd prefer